tests/forms_tests/tests/test_forms.py

```diff
@@ -3014,6 +3014,134 @@ class UserRegistration(Form):
+                'Password:</label><input type="password" name="password" '
+                'required aria-describedby="id_password_helptext" id="id_password">'
+                '<span class="helptext" id="id_password_helptext">'
+                "Wählen Sie mit Bedacht.</span></div>",
+            )
+            self.assertHTMLEqual(
+                p.as_ul(),
+                '<li><label for="id_username">Username:</label>'
+                '<input type="text" name="username" maxlength="10" required '
+                'aria-describedby="id_username_helptext" id="id_username">'
+                '<span class="helptext" id="id_username_helptext">'
+                'e.g., user@example.com</span></li><li><label for="id_password">'
+                'Password:</label><input type="password" name="password" '
+                'required aria-describedby="id_password_helptext" id="id_password">'
+                '<span class="helptext" id="id_password_helptext">'
+                "Wählen Sie mit Bedacht.</span></li>",
+            )
+            self.assertHTMLEqual(
+                p.as_p(),
+                '<p><label for="id_username">Username:</label>'
+                '<input type="text" name="username" maxlength="10" required '
+                'aria-describedby="id_username_helptext" id="id_username">'
+                '<span class="helptext" id="id_username_helptext">'
+                'e.g., user@example.com</span></p><p><label for="id_password">'
+                'Password:</label><input type="password" name="password" '
+                'required aria-describedby="id_password_helptext" id="id_password">'
+                '<span class="helptext" id="id_password_helptext">'
+                "Wählen Sie mit Bedacht.</span></p>",
+            )
+            self.assertHTMLEqual(
+                p.as_table(),
+                '<tr><th><label for="id_username">Username:</label></th><td>'
+                '<input type="text" name="username" maxlength="10" required '
+                'aria-describedby="id_username_helptext" id="id_username"><br>'
+                '<span class="helptext" id="id_username_helptext">'
+                "e.g., user@example.com</span></td></tr><tr><th><label "
+                'for="id_password">Password:</label></th><td><input type="password" '
+                'name="password" required aria-describedby="id_password_helptext" '
+                'id="id_password"><br><span class="helptext" id="id_password_helptext">'
+                "Wählen Sie mit Bedacht.</span></td></tr>",
+            )
+
+    def test_custom_describedby(self):
+        """aria-describedby provided to the widget overrides the default"""
+
+        class UserRegistration(Form):
+            username = CharField(
+                max_length=10,
+                help_text="e.g., user@example.com",
+                widget=TextInput(attrs={"aria-describedby": "custom-description"}),
```

Comment:
I'm not familiar with `aria-describedby` but from previous comments, it seemed that if a user sets the attribute, the like in this test, then the screen readers would not be able to match the help_text with the field. If my understanding is correct, I think we'd need to consider having a more robust `aria-describedby` customization story. Was this discussed already?

Reply:
We did discuss this, see [#15246#discussion_r777873099 and onwards](https://github.com/django/django/pull/15246#discussion_r777873099), and all of [#16185](https://github.com/django/django/pull/16185). To answer your question – screen readers will still point to the help text, _if the customisation includes it in the `aria-describedby`_. So this test would be clearer if it set the attribute to `custom-description id_username_helptext` as we recommend in the docs.

---

In Django code we can’t tell what order the descriptions should go in, because we don’t know whether the custom description is visually before or after the help text. Hence the documentation mentioning any usage of the attribute needs to include a reference to the help text element (if used), in the appropriate order, which is to be determined by the person setting this custom value. The reasoning for using `aria-describedby` customisations as-is is:

- It’s more predictable for everyone
- It allows customisations to have the final say on what the appropriate order is (or do something more advanced potentially)

The alternative would be another mechanism that allows explicitly ordering different references.

Note technically speaking this doesn’t fix all of [#32819](https://code.djangoproject.com/ticket/32819) (neither did #15246 or #16185), we’ll still need the `aria-describedby` to include a reference to any error messages. The current work still feels like a good step forward though.

Reply:
Thanks @thibaudcolas for the explanation. I agree this definitely feels like a step forward.

Regarding my question, thank you for the references, I read them. My first expectation was the following (which I don't think it was proposed on the already discussed options?):

* if no `aria-describedby` is given, use the correct one (this is non controversial)
* elif `aria-describedby` includes the expected help text id, use as is (also non controversial I think)
* else, append? prepend? the expected help text id to ensure the whole thing makes sense.

Pseudocode:

```python
        helptext_id = f"{self.id_for_label}_helptext"
        if custom_aria_described_by_id := widget.attrs.get("aria-describedby"):
            if helptext_id not in custom_aria_described_by_id:
                custom_aria_described_by_id += f" {helptext_id}"
            attrs["aria-describedby"] = custom_aria_described_by_id
        elif self.field.help_text and self.id_for_label:
            attrs["aria-describedby"] = helptext_id
```

Is it too simplistic on my part to propose this? :-)

Reply:
I think that if folk have set a custom `aria-describedby` then we should just use that and that users are on their own at this point. 

If we always add Django's reference we may be adding asking the screen reader to read something that doesn't actually exist. 

Here's a real example. With your proposal it would add two `aria-describedby` labels to the input, but the one added by Django would be incorrect. 

https://github.com/wildfish/crispy-forms-gds/blob/master/tests/layout/results/radios/choices.html

(p.s. if you did try and checkout that repo to see it in action, you'll need to pin `django-crispy-forms<2.0`

Reply:
In that "else, append? prepend?" scenario, we’re still in a situation where Django has to decide which order is "correct" without any information, so there’s a good 50% chance we’d get it wrong. I’d assume it’s better for everyone if `aria-describedby` is left for form implementers to set correctly, so there’s less magic involved.

If we’re ok with some magic, then I think I’d recommend something like this that allows implementers to decide if and how Django’s help text is included:

```python
attrs["aria-describedby"] = custom_aria_described_by_id.replace("django-help_text", helptext_id)
```

…so we can tell implementers "to include Django’s help text, place `django-help_text` in your `aria-describedby` definition before or after any other description as appropriate. Django will then replace this with the field’s help text id when rendering the field".

Reply:
Thank you everyone. With the provided context I agree we should not change the value of the custom `aria-describedby`, if given. I would not use a placeholder like `django-help_text`, I think it could lead to confusion and/or false expectations around the apparent magic.